### ISSUE TITLE
Map to Interface

### ISSUE THREAD

**Issue by bslbckr** [agileobjects/AgileMapper#123]
We are currently evaluating AgileMapper and are facing some problems mapping from a protobuf message to an interface.

Protobuf does not support inheritance and thus we have to find a way around it, which we currently solve using protobuf's `oneof`-construct. This yields the following c#-classes generated from protobuf:
```
public class CompositeDTO {
  CompositeDTOInstanceCase InstanceCase { public get: }
  GroupDTO { public get: public set: }
  LeafDTO { public get; public set; }
}

public class GroupDTO {
  // RepeatedField inherits from IList
  public RepeatedField<CompositeDTO> Children { public get; } = new RepeatedField ...
}

public class LeafDTO {}
```
We want to map them to the following objects in our domain-model:
```
public interface IComposite {}
public class Group : IComposite {
  IList<IComposite> Children { public get; } = new List<IComposite>();
}
public class Leaf : IComposite {}
```
Obviously, we want AgileMapper to map `GroupDTO` to `Group` and `LeafDTO` to `Leaf`. However, so far we fail to give the correct mapping configuration. We tried conditional mapping configurations using `if((d) => d.CompositeDTOInstanceCase ==...).MapTo<Group>().And.If((d) => ...).MapTo<Leaf>()`, or `From<CompositeDTO>().To<IComposite>().CreateInstancesUsing(...)` without success. In all situations the `Children` were `null`, whereas the first / outer mapping from `GroupDTO` to `Group` which was triggered by our call to `IMapper` looks ok; i.e. `mapper.Map(myGroupDTO).ToANew<Group>()`.

What we can tell from looking at the execution plan is, that the `null` values come from AgileMapper's assumption, that it cannot create any instances for `IComposite` and thus returns null.

Is ther anything in the configuration we are overlooking at the moment or are we doing things fundamentally wrong? Any help is appreciated.

**Comment by bslbckr** [comment]
I created a little .NetFiddle to illustrate the problems we have and give a MWE: https://dotnetfiddle.net/VQmiRt

**Comment by SteveWilkes** [comment]
Hi!

Thanks very much for the feedback and the fiddle - your scenario should definitely be supported, and there's one or two bugs here I'll fix.

The `GroupDTO` -> `Group` mapping plan shows the `GroupDTO.Children` being projected to map the `CompositeDTO.Leaf` objects to `Group.Children`:

```C#
var leafDTOs = gdtoToGData.Source.Children.Project(cdto => cdto.Leaf);
```

...the subsequent element mapping is then `LeafDTO` -> `IComposite`, which doesn't match the configuration. It shouldn't be performing that projection (it thinks `CompositeDTO` is a linking object), so I'll sort that out.

The mapper's also incorrectly giving up when asked to map to an interface - the `.MapTo<>` configurations should cover it.

Cheers!

Steve

**Comment by SteveWilkes** [comment]
This is fixed in v1.3-preview2, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.3.0-preview2).

- I've improved the join entity detection, so `Children` is no longer projected to a collection of `Leaf` objects

- `.MapTo<T>` now supports interfaces as you'd expect

- The configuration to set up the `CompositeDTO` -> `IComposite` mapping is:

```C#
mapper.WhenMapping
    .From<CompositeDTO>().To<IComposite>()
    .If(data => data.Source.CompositeDTOInstanceCase == CompositeDTO.InstanceCase.Group)
    .MapTo<Group>()
    .And
    .If(data => data.Source.CompositeDTOInstanceCase == CompositeDTO.InstanceCase.Leaf)
    .MapTo<Leaf>();
```

...which sets up `CompositeDTO` -> `Leaf` and `CompositeDTO` -> `Group` mappings.

Then, to configure which `CompositeDTO` member to use in those mappings:

```C#
mapper.WhenMapping
    .From<CompositeDTO>().To<Group>()
    .Map(data => data.Source.Group)
    .ToTarget();

mapper.WhenMapping
    .From<CompositeDTO>().To<Leaf>()
    .Map(data => data.Source.Leaf)
    .ToTarget();
```

Finally, I've added a typed `.ToTarget()` overload so we can configure the mapping more simply:

```C#
mapper.WhenMapping
    .From<CompositeDTO>().To<IComposite>()
    .If(data => data.Source.CompositeDTOInstanceCase == CompositeDTO.InstanceCase.Group)
    .Map(data => data.Source.Group)
    .ToTarget<Group>();

mapper.WhenMapping
    .From<CompositeDTO>().To<IComposite>()
    .If(data => data.Source.CompositeDTOInstanceCase == CompositeDTO.InstanceCase.Leaf)
    .Map(data => data.Source.Leaf)
    .ToTarget<Leaf>();
```

I've added [a new fiddle](https://dotnetfiddle.net/aLe1Rk) showing it all working. Thanks again for the feedback, and let me know if you encounter any more issues! :)

**Comment by bslbckr** [comment]
> This is fixed in v1.3-preview2, which is now available on NuGet
> <https://www.nuget.org/packages/AgileObjects.AgileMapper/1.3.0-preview2>.
>
Thanks, I'll give the preview a try.

> I've added a new fiddle <https://dotnetfiddle.net/aLe1Rk> showing it all
> working. Thanks again for the feedback, and let me know if you encounter
> any more issues! :)
>

I will :)
Thank you very much for your quick and supportive response!

**Comment by SteveWilkes** [comment]
This is fixed in v1.3, now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/).

Thanks again! :)
